src/main/java/graphql/execution/FieldDirectives.java

```diff
@@ -0,0 +1,179 @@
+    private final Map<Field, Map<String, GraphQLDirective>> fragmentDefDirectives;
+    private final Map<String, GraphQLDirective> operationDirectives;
+
+    @Internal
+    public FieldDirectives(Document document, GraphQLSchema schema, Map<String, FragmentDefinition> fragmentsByName, Map<String, Object> variables, OperationDefinition operationDefinition, DirectivesResolver directivesResolver) {
+        DirectiveCollector collector = DirectiveCollector.collect(document, schema, fragmentsByName, variables, operationDefinition.getName(), directivesResolver);
+        this.fieldDirectives = collector.getFieldDirectives();
+        this.inlineFragmentDirectives = collector.getInlineFragmentDirectives();
+        this.fragmentDirectives = collector.getFragmentDirectives();
+        this.fragmentDefDirectives = collector.getFragmentDefDirectives();
+        this.operationDirectives = directivesResolver.resolveDirectives(operationDefinition.getDirectives(), schema, variables);
+    }
+
+    public Map<String, GraphQLDirective> getFieldDirectives(Field field) {
+        return fieldDirectives.get(field);
+    }
+
+    public Map<String, List<GraphQLDirective>> getFieldDirectives(List<Field> fields) {
+        return getFieldDirectives(fieldDirectives, fields);
+    }
+
+    public Map<String, List<GraphQLDirective>> getInlineFragmentDirectives(List<Field> fields) {
+        return getFieldDirectives(inlineFragmentDirectives, fields);
+    }
+
+    public Map<String, List<GraphQLDirective>> getFragmentDirectives(List<Field> fields) {
+        return getFieldDirectives(fragmentDirectives, fields);
+    }
+
+    public Map<String, List<GraphQLDirective>> getFragmentDefDirectives(List<Field> fields) {
+        return getFieldDirectives(fragmentDefDirectives, fields);
+    }
+
+    public Map<String, GraphQLDirective> getOperationDirectives() {
+        return operationDirectives;
+    }
+
+    private Map<String, List<GraphQLDirective>> getFieldDirectives(Map<Field, Map<String, GraphQLDirective>> directivesMap, List<Field> fields) {
+        return fields.stream()
+                .flatMap(field -> directivesMap.getOrDefault(field, Collections.emptyMap()).values().stream())
+                .collect(Collectors.groupingBy(GraphQLDirective::getName));
+    }
+
+    private static class DirectiveCollector extends QueryVisitorStub {
+
+        private final GraphQLSchema schema;
+        private final Map<String, Object> variables;
+        private final DirectivesResolver directivesResolver;
+        private final Map<Field, Map<String, GraphQLDirective>> fieldDirectives;
+        private final Map<Field, Map<String, GraphQLDirective>> inlineFragmentDirectives;
+        private final Map<Field, Map<String, GraphQLDirective>> fragmentDirectives;
+        private final Map<Field, Map<String, GraphQLDirective>> fragmentDefDirectives;
+        private final Map<Field, Set<Node>> fragmentsPerField;
```

Comment:
Hmmm - can Field be a good key?  It has not equals / hasCode.  This collects all fields in the Document.

But later when we match them up we need to do it by Field instance.

But Field objects can be "rewritten" and I think this is going to do "identity" matching.  And we may have changed a field by execution time

Hmmm - I think this is a non starter in terms of being able to find all directives "eagerly" before document execution.

I think we are going to have to collect them from a DataFetcher as is.

In the tests can we prove that we can have repeated field names in lower levels are still work as expected

Reply:
cc: @andimarek ^